File: chatbot-service/chatbot.py
import os
import requests
from dotenv import load_dotenv
from langdetect import detect

load_dotenv()

GROQ_API_KEY = os.getenv("GROQ_API_KEY")

print(f"[Chatbot Init] Loaded GROQ_API_KEY: {'✓' if GROQ_API_KEY else '✗ MISSING'}")

def chatbot_response(user_message):
    if not GROQ_API_KEY:
        return "Error: GROQ_API_KEY not configured. Please set it in .env file."

    # 🔍 Detect language
    try:
        lang = detect(user_message)
    except:
        lang = "en"

    # 🎯 Map language
    if lang == "mr":
        language = "Marathi"
        script_rule = "Respond ONLY in Marathi using Devanagari script (मराठी लिपी). Do NOT use Roman Marathi."
    elif lang == "hi":
        language = "Hindi"
        script_rule = "Respond ONLY in Hindi using Devanagari script. Do NOT mix English."
    else:
        language = "English"
        script_rule = "Respond ONLY in English."

    url = "https://api.groq.com/openai/v1/chat/completions"

    headers = {
        "Authorization": f"Bearer {GROQ_API_KEY}",
        "Content-Type": "application/json"
    }

    system_prompt = f"""
You are a warm, friendly, and highly knowledgeable Ayurvedic wellness assistant for the AyurSutra platform.

YOUR PERSONALITY:
- Speak like a caring Ayurvedic doctor or wellness guide
- Be calm, friendly, empathetic, and supportive
- Make the user feel heard and comfortable
- Avoid robotic or overly technical responses

CORE EXPERTISE:
1. Tridosha Theory: Vata (वात), Pitta (पित्त), Kapha (कफ)
2. Panchakarma Therapies: Vamana, Virechana, Basti, Nasya, Raktamokshana
3. Herbal Remedies: Triphala, Chyawanprash, Dashamoola
4. Diet & Lifestyle: Dinacharya, Ritucharya
5. Yoga & Pranayama

STRICT LANGUAGE RULES:
- {script_rule}
- Do NOT mix languages
- Do NOT translate
- Output must be in ONE language only

DOMAIN RULES:
- Only answer Ayurveda, Panchakarma, herbs, lifestyle, and wellness
- Keep explanations simple, practical, and correct
- Avoid modern medical diagnosis unless necessary

CONVERSATION STYLE (VERY IMPORTANT):
- Always start with a warm, natural tone
- Keep answers concise (2–4 sentences)
- After answering, ask 1 relevant follow-up question to understand the user better
- Personalize responses based on user inputs (symptoms, lifestyle, habits)
- If user shares symptoms, gently explore:
  • duration
  • diet habits
  • sleep pattern
  • stress level

FOLLOW-UP QUESTION EXAMPLES:
- "Can you tell me when this started?"
- "How is your digestion and appetite?"
- "Do you feel more heat or cold in your body?"
- "How is your sleep these days?"

GUIDANCE STYLE:
- Suggest small, actionable steps (diet, herbs, routines)
- Prefer natural remedies over complex treatments initially
- For Panchakarma, suggest consulting a practitioner if needed

SAFETY:
- Do not give extreme or risky medical advice
- If condition seems serious, gently suggest consulting a doctor

IF OUTSIDE DOMAIN:
Reply in {language}:
"I specialize in Ayurveda and wellness. I’d be happy to help you with your health, diet, or lifestyle concerns."
"""

    data = {
        "model": "llama-3.1-8b-instant",
        "messages": [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_message}
        ],
        "max_tokens": 120
    }

    try:
        print(f"[API Call] Sending request to Groq API for language: {language}")
        response = requests.post(url, headers=headers, json=data, timeout=30)
        print(f"[API Response] Status: {response.status_code}")
        
        result = response.json()

        if response.status_code != 200:
            error_msg = result.get("error", {}).get("message", "Unknown error")
            print(f"[API Error] {error_msg}")
            return f"Error: {error_msg}"

        if "choices" not in result:
            print(f"[API Error] No choices in response: {result}")
            return "Error: Invalid response format from AI service"

        reply = result["choices"][0]["message"]["content"]

        # 🔥 CLEANUP (remove parenthesized translations/notes but keep the main answer)
        # e.g. "Vata (वात) is ..." => "Vata is ..."
        try:
            import re
            reply = re.sub(r"\s*\([^)]*\)", "", reply)
        except Exception:
            pass

        print(f"[API Success] Generated response ({len(reply)} chars)")
        return reply.strip()

    except requests.exceptions.Timeout:
        error = "Error: API request timed out (30s)"
        print(f"[API Timeout] {error}")
        return error
    except requests.exceptions.ConnectionError:
        error = "Error: Cannot connect to Groq API. Check internet connection."
        print(f"[API Connection Error] {error}")
        return error
    except Exception as e:
        error = f"Error: {str(e)}"
        print(f"[API Exception] {error}")
        return error

File: server/src/routes/chatbot.js
import express from "express";
import Groq from "groq-sdk";

const router = express.Router();
const groq = new Groq({ apiKey: process.env.GROQ_API_KEY });

const OUT_OF_SCOPE_REPLY_EN =
    "I specialize in Ayurveda and wellness. Please ask a question related to Ayurveda, Panchakarma, herbs, diet, dosha, yoga, or lifestyle healing.";
const OUT_OF_SCOPE_REPLY_HI =
    "मैं आयुर्वेद और वेलनेस से जुड़े प्रश्नों में विशेषज्ञ हूं। कृपया आयुर्वेद, पंचकर्म, जड़ी-बूटियां, आहार, दोष, योग या जीवनशैली उपचार से जुड़ा प्रश्न पूछें।";

const AYURVEDA_KEYWORDS = [
    "ayurveda",
    "ayurvedic",
    "panchakarma",
    "dosha",
    "vata",
    "pitta",
    "kapha",
    "agni",
    "ama",
    "dinacharya",
    "ritucharya",
    "rasayana",
    "abhyanga",
    "nasya",
    "basti",
    "vamana",
    "virechana",
    "triphala",
    "ashwagandha",
    "chyawanprash",
    "giloy",
    "tulsi",
    "herb",
    "herbal",
    "pranayama",
    "yoga",
    "meditation",
    "wellness",
    "digestion",
    "constipation",
    "acidity",
    "sleep",
    "stress",
    "anxiety",
    "immunity",
];

function normalizeText(value = "") {
    return String(value).toLowerCase().trim();
}

function isAyurvedaQuery(message = "") {
    const text = normalizeText(message);
    if (!text) return false;
    return AYURVEDA_KEYWORDS.some((k) => text.includes(k));
}

function normalizeHistory(history = []) {
    if (!Array.isArray(history)) return [];
    return history
        .map((m) => ({
            role: m?.role === "assistant" ? "assistant" : "user",
            content: safeText(m?.content).trim(),
        }))
        .filter((m) => m.content);
}

function looksLikeFollowup(message = "") {
    const text = normalizeText(message);
    if (!text) return false;
    // Typical short follow-up prompts that rely on prior context.
    const followupSignals = [
        "what about",
        "and this",
        "is it safe",
        "safe?",
        "how long",
        "how often",
        "dosage",
        "dose",
        "when should",
        "can i",
        "should i",
        "then what",
        "next step",
        "why",
        "how",
        "ok then",
        "what now",
        "in hindi",
        "in marathi",
    ];
    if (text.length <= 42) return true;
    return followupSignals.some((s) => text.includes(s));
}

function hasRecentAyurvedaContext(history = []) {
    const normalized = normalizeHistory(history);
    // Look at recent user+assistant turns to decide if chat is currently in domain.
    const recent = normalized.slice(-6);
    return recent.some((m) => isAyurvedaQuery(m.content));
}

function safeText(v, fallback = "") {
    return typeof v === "string" ? v : fallback;
}

function normalizeLocale(value = "") {
    const v = normalizeText(value);
    return v.startsWith("hi") ? "hi" : "en";
}

function normalizeStructuredReply(raw) {
    const obj = raw && typeof raw === "object" ? raw : {};
    const title = safeText(obj.title, "AyurVaidya Guidance");
    const summary = safeText(obj.summary, "");
    const tip = safeText(obj.tip, "");
    const sectionsRaw = Array.isArray(obj.sections) ? obj.sections : [];
    const sections = sectionsRaw
        .map((s) => {
            if (!s || typeof s !== "object") return null;
            const type = safeText(s.type);
            const heading = safeText(s.heading);
            if (type === "text") {
                return { type, heading, content: safeText(s.content) };
            }
            if (type === "list" || type === "steps") {
                const items = Array.isArray(s.items) ? s.items.map((x) => safeText(x)).filter(Boolean) : [];
                return { type, heading, items };
            }
            if (type === "table") {
                const rows = Array.isArray(s.rows)
                    ? s.rows
                          .map((r) => ({
                              label: safeText(r?.label),
                              value: safeText(r?.value),
                          }))
                          .filter((r) => r.label || r.value)
                    : [];
                return { type, heading, rows };
            }
            return null;
        })
        .filter(Boolean);
    return { title, summary, sections, tip };
}

// ─── Chat with Digital Vaidya (Ayurvedic AI Assistant) ───
router.post("/chat", async (req, res) => {
    try {
        const { message, conversationHistory = [], userProfile, locale } = req.body;
        const lang = normalizeLocale(locale);
        const isHindi = lang === "hi";
        const userMessage = safeText(message).trim();
        if (!userMessage) {
            return res.status(400).json({ success: false, error: "message is required" });
        }

        const history = normalizeHistory(conversationHistory);
        const inDomainNow = isAyurvedaQuery(userMessage);
        const contextFollowup = looksLikeFollowup(userMessage) && hasRecentAyurvedaContext(history);

        // Context-aware domain guard:
        // - allow explicit Ayurveda queries
        // - allow short follow-ups if recent context is already Ayurveda
        if (!inDomainNow && !contextFollowup) {
            return res.json({
                success: true,
                inScope: false,
                reply: isHindi ? OUT_OF_SCOPE_REPLY_HI : OUT_OF_SCOPE_REPLY_EN,
                structured: {
                    title: isHindi ? "आयुर्वैद्य दायरा सूचना" : "AyurVaidya Scope Notice",
                    summary: isHindi ? OUT_OF_SCOPE_REPLY_HI : OUT_OF_SCOPE_REPLY_EN,
                    sections: [],
                    tip: isHindi
                        ? "दोष संतुलन, आहार, जड़ी-बूटियों, पंचकर्म, योग, पाचन, नींद या तनाव के बारे में पूछें।"
                        : "Try asking about dosha balance, diet, herbs, Panchakarma, yoga, digestion, sleep, or stress.",
                },
            });
        }

        const systemPrompt = `You are "AyurVaidya", a knowledgeable and compassionate AI Ayurvedic health assistant for AyurSutra platform. You are trained on classical Ayurvedic texts and provide authentic guidance.

**Classical Text Knowledge (Charaka Samhita & Sushruta Samhita):**
- Charaka Samhita: The foundational text of Ayurveda covering 8 branches (Ashtanga Ayurveda). Key concepts: Tridosha Siddhanta, Saptadhatu, Panchamahabhoota, Dinacharya, Ritucharya, Sadvritta.
- Sushruta Samhita: The surgical text of Ayurveda. Key references: Shalya Tantra, Marma points, wound healing, Panchakarma detoxification procedures.
- Ashtanga Hridaya: Vagbhata's compilation combining both Charaka and Sushruta traditions.

**Core Expertise:**
1. **Tridosha Theory**: Vata (वात), Pitta (पित्त), Kapha (कफ) — assessment, imbalance detection, and balancing techniques
2. **Panchakarma Therapies**: Vamana (वमन), Virechana (विरेचन), Basti (बस्ति), Nasya (नस्य), Raktamokshana (रक्तमोक्षण) — as per Charaka Samhita Siddhisthana
3. **Herbal Remedies (Dravyaguna Shastra)**: Classical formulations like Triphala, Chyawanprash, Dashamoola, Mahanarayan Taila
4. **Diet & Lifestyle (Ahara & Vihara)**: Dinacharya (दिनचर्या - daily routines), Ritucharya (ऋतुचर्या - seasonal routines), Pathya-Apathya (dietary do's & don'ts)
5. **Yoga & Pranayama**: Therapeutic sequences per Patanjali Yoga Sutras

**Multilingual Support:**
- You understand and can respond with Hindi (हिंदी) and Sanskrit (संस्कृत) Ayurvedic terms with translations
- When using Ayurvedic terms, always provide: Sanskrit name → Hindi equivalent → English translation
- Example: "Agni (अग्नि) refers to the digestive fire that governs metabolism"
- You can interpret prescriptions written in Hindi, Sanskrit, or transliterated forms

${userProfile ? `Patient Profile:
- Name: ${userProfile.name}
- Dosha: ${userProfile.dosha || "Not assessed"}
- Health Score: ${userProfile.healthScore || "Not assessed"}` : ""}

Guidelines:
- Provide compassionate, evidence-informed Ayurvedic guidance rooted in classical texts
- Reference specific classical texts when giving recommendations (e.g., "As described in Charaka Samhita, Chikitsasthana Ch.15...")
- Always recommend consulting a qualified Vaidya (Ayurvedic doctor) for serious conditions
- Keep responses concise and practical
- Include practical tips the user can act on immediately
- Use simple language, explain Sanskrit/Ayurvedic terms when used
- Support multilingual queries in Hindi and English
- IMPORTANT: Respond in ${isHindi ? "Hindi (hi-IN)" : "English (en-IN)"} for this user unless explicitly asked to switch language.
- NEVER provide emergency medical advice — direct to emergency services if needed

VERY IMPORTANT OUTPUT FORMAT:
Return ONLY valid JSON in this exact schema:
{
  "title": "<short title>",
  "summary": "<1-2 sentence quick answer>",
  "sections": [
    { "heading": "Why this happens", "type": "text", "content": "<brief explanation>" },
    { "heading": "What to do now", "type": "steps", "items": ["<step 1>", "<step 2>", "<step 3>"] },
    { "heading": "Helpful options", "type": "list", "items": ["<item 1>", "<item 2>"] }
  ],
  "tip": "<single practical tip>"
}
Allowed section types are only: "text", "list", "table", "steps".
Do not include markdown. Do not include any text outside JSON.`;

        const messages = [
            { role: "system", content: systemPrompt },
            ...history.slice(-8),
            { role: "user", content: userMessage },
        ];

        const completion = await groq.chat.completions.create({
            messages,
            model: "llama-3.3-70b-versatile",
            temperature: 0.3,
            max_tokens: 600,
            response_format: { type: "json_object" },
        });

        const rawText = completion.choices[0]?.message?.content || "{}";
        let structuredRaw = {};
        try {
            structuredRaw = JSON.parse(rawText);
        } catch {
            structuredRaw = {
                title: isHindi ? "आयुर्वैद्य मार्गदर्शन" : "AyurVaidya Guidance",
                summary: safeText(
                    rawText,
                    isHindi
                        ? "मैं आयुर्वेद और वेलनेस से जुड़े प्रश्नों में मदद कर सकता हूं।"
                        : "I can help with Ayurveda and wellness questions."
                ),
                sections: [],
                tip: "",
            };
        }
        const structured = normalizeStructuredReply(structuredRaw);
        const plainReply = [structured.summary, structured.tip ? `${isHindi ? "सुझाव" : "Tip"}: ${structured.tip}` : ""]
            .filter(Boolean)
            .join("\n\n");

        res.json({
            success: true,
            inScope: true,
            reply: plainReply,
            structured,
        });
    } catch (error) {
        console.error("Chatbot error:", error);
        res.status(500).json({
            success: false,
            error: "Failed to get response from AyurVaidya",
        });
    }
});

export default router;
